lib/queueservice.js

```diff
@@ -0,0 +1,82 @@
+import xml2js from 'xml2js';
+import Debug from 'debug';
+import request from 'superagent-promise';
+
+let debug = Debug('taskcluster-docker-worker:queueService');
+
+export default class TaskQueue {
+  constructor(config) {
+    this.queues;
+    this.workerType = config.workerType;
+    this.provisionerId = config.provisionerId;
+    this.client = config.queue;
+    this.log = config.log;
+  }
+
+  async getQueues() {
+    if (!this.queues || (new Date(this.queues.expires)).getTime() > Date.now()) {
+      this.queues = await this.client.pollTaskUrls(this.provisionerId, this.workerType);
+    }
+    return this.queues.queues;
+  }
+
+  async getTasksFromQueue(queue, numberOfTasks) {
```

Comment:
How about just passing in an integer instead of the actual queue?
Then the list of queue objects shouldn't have the option of leaving this class. Abstracting that logic away would be neat.

Also then you can do the check of signed url expiration and potential call to `pollTaskUrls` before polling the azure queue, a head of each and every poll...
Note, this way you might accept that it only has to be 3 min a head of expiration. Before refreshing, and chance of making a request with expired signature is much lower.

Note, you should handle the special case where the integer given is higher than the number of queues, by doing integer modulo number of queues. This could happen if after refreshing you get fewer azure queues than before.

Note in this setup, all you return to `TaskListener` from `getQueues` is the number of queues that exists. Granted you might want to change `getQueues` into `getNumberOfQueues`.